File: cpp/src/analytics/drawdown.cpp
/**
 * AlgoForge — src/analytics/drawdown.cpp
 * Drawdown series, summary, and episodes.
 */

#include "core/analytics.hpp"

#include <algorithm>
#include <cmath>
#include <span>
#include <vector>

namespace algoforge::analytics {

DrawdownSummary compute_drawdown(std::span<const double> equity) noexcept {
    DrawdownSummary ds;
    const int n = static_cast<int>(equity.size());
    if (n == 0) return ds;

    ds.dd_pct_series.resize(n);
    ds.dd_abs_series.resize(n);

    double peak     = equity[0];
    double max_pct  = 0.0;
    double max_abs  = 0.0;

    for (int i = 0; i < n; ++i) {
        if (equity[i] > peak) peak = equity[i];
        double dd_abs = equity[i] - peak;
        double dd_pct = (peak > 1e-15) ? (dd_abs / peak * 100.0) : 0.0;
        ds.dd_pct_series[i] = dd_pct;
        ds.dd_abs_series[i] = dd_abs;
        if (dd_pct < max_pct) max_pct = dd_pct;
        if (dd_abs < max_abs) max_abs = dd_abs;
    }
    ds.max_dd_pct = max_pct;
    ds.max_dd_abs = max_abs;

    /* ── Drawdown episodes: contiguous runs of dd_pct < 0 ── */
    int i = 0;
    while (i < n) {
        /* find episode start: first bar below prior peak */
        if (ds.dd_pct_series[i] < 0.0) {
            DrawdownEpisode ep;
            ep.start_idx   = i;
            ep.trough_idx  = i;

            /* find trough */
            while (i < n && ds.dd_pct_series[i] < 0.0) {
                if (ds.dd_pct_series[i] < ds.dd_pct_series[ep.trough_idx]) {
                    ep.trough_idx = i;
                }
                ++i;
            }
            /* i now points to recovery or end */

            ep.depth_pct = ds.dd_pct_series[ep.trough_idx];
            ep.depth_abs = ds.dd_abs_series[ep.trough_idx];

            /* find recovery: first index where equity >= peak before episode */
            /* The peak before the episode is the value at (start_idx - 1)
               or equity[0] if start_idx == 0 */
            double pre_peak = (ep.start_idx > 0) ? equity[ep.start_idx - 1] : equity[0];

            ep.recover_idx   = -1;
            ep.recovery_bars = -1;
            for (int j = ep.trough_idx + 1; j < n; ++j) {
                if (equity[j] >= pre_peak) {
                    ep.recover_idx = j;
                    break;
                }
            }

            if (ep.recover_idx >= 0) {
                ep.duration_bars = ep.recover_idx - ep.start_idx;
                ep.recovery_bars = ep.recover_idx - ep.trough_idx;
            } else {
                ep.duration_bars = n - ep.start_idx;
                ep.recovery_bars = -1;
            }

            ds.episodes.push_back(ep);
        } else {
            ++i;
        }
    }

    return ds;
}

std::vector<DrawdownEpisode> top5_drawdowns(std::span<const double> equity) noexcept {
    auto ds = compute_drawdown(equity);
    auto& eps = ds.episodes;

    /* sort by depth_pct ascending (most negative = deepest first) */
    std::sort(eps.begin(), eps.end(), [](const DrawdownEpisode& a, const DrawdownEpisode& b) {
        return a.depth_pct < b.depth_pct;
    });

    if (eps.size() > 5) eps.resize(5);
    return eps;
}

} /* namespace algoforge::analytics */

File: js/src/analytics/drawdown.js
/**
 * drawdown.js — drawdown series, episodes, and top-5 summary.
 * Spec §4 (Drawdown section).
 */

/**
 * Compute peak array and drawdown series from equity curve.
 * @param {number[]} eq  equity curve, strictly positive values expected
 * @returns {{
 *   peaks:    number[],
 *   ddPct:    number[],
 *   ddAbs:    number[],
 *   maxDdPct: number,
 *   maxDdAbs: number,
 * }}
 */
export function drawdownSeries(eq) {
    const n = eq.length;
    if (n === 0) {
        return { peaks: [], ddPct: [], ddAbs: [], maxDdPct: 0, maxDdAbs: 0 };
    }
    const peaks = new Array(n);
    const ddPct = new Array(n);
    const ddAbs = new Array(n);
    let peak = eq[0];
    let maxDdPct = 0;
    let maxDdAbs = 0;

    for (let i = 0; i < n; i++) {
        if (eq[i] > peak) peak = eq[i];
        peaks[i] = peak;
        const pct = peak === 0 ? 0 : (eq[i] - peak) / peak * 100;
        const abs = eq[i] - peak;
        ddPct[i] = pct;
        ddAbs[i] = abs;
        if (pct < maxDdPct) maxDdPct = pct;
        if (abs < maxDdAbs) maxDdAbs = abs;
    }

    return { peaks, ddPct, ddAbs, maxDdPct, maxDdAbs };
}

/**
 * Compute drawdown episodes: contiguous runs of ddPct < 0.
 * Each episode: { startIdx, troughIdx, recoverIdx, depthPct, depthAbs, durationBars, recoveryBars }
 * recoverIdx = -1 if not recovered by end of series.
 * @param {number[]} eq
 * @returns {Array<{startIdx:number,troughIdx:number,recoverIdx:number,depthPct:number,depthAbs:number,durationBars:number,recoveryBars:number}>}
 */
export function drawdownEpisodes(eq) {
    const n = eq.length;
    if (n === 0) return [];
    const { peaks, ddPct, ddAbs } = drawdownSeries(eq);

    const episodes = [];
    let inDd = false;
    let startIdx = -1;
    let troughIdx = -1;
    let depthPct = 0;
    let depthAbs = 0;

    for (let i = 0; i < n; i++) {
        if (!inDd) {
            if (ddPct[i] < 0) {
                // Start of new episode
                inDd = true;
                startIdx = i;
                troughIdx = i;
                depthPct = ddPct[i];
                depthAbs = ddAbs[i];
            }
        } else {
            // Inside drawdown
            if (ddPct[i] < depthPct) {
                depthPct = ddPct[i];
                depthAbs = ddAbs[i];
                troughIdx = i;
            }
            if (ddPct[i] === 0) {
                // Recovered
                const durationBars = i - startIdx;
                const recoveryBars = i - troughIdx;
                episodes.push({
                    startIdx,
                    troughIdx,
                    recoverIdx: i,
                    depthPct,
                    depthAbs,
                    durationBars,
                    recoveryBars,
                });
                inDd = false;
                startIdx = -1;
                troughIdx = -1;
                depthPct = 0;
                depthAbs = 0;
            }
        }
    }

    // Handle open episode at end of series
    if (inDd) {
        const durationBars = (n - 1) - startIdx;
        episodes.push({
            startIdx,
            troughIdx,
            recoverIdx: -1,
            depthPct,
            depthAbs,
            durationBars,
            recoveryBars: -1,
        });
    }

    return episodes;
}

/**
 * Return top-5 drawdown episodes sorted by depthPct ascending (deepest first).
 * @param {number[]} eq
 * @returns {Array}
 */
export function top5Drawdowns(eq) {
    const eps = drawdownEpisodes(eq);
    eps.sort((a, b) => a.depthPct - b.depthPct);
    return eps.slice(0, 5);
}
